src/coreWebVitals.ts

```diff
@@ -0,0 +1,121 @@
+import type { ReportHandler } from 'web-vitals';
+import { getCLS, getFCP, getFID, getLCP, getTTFB } from 'web-vitals';
+
+type CoreWebVitalsPayload = {
+	page_view_id: string | null;
+	browser_id: string | null;
+	fid: null | number;
+	cls: null | number;
+	lcp: null | number;
+	fcp: null | number;
+	ttfb: null | number;
+};
+
+const coreWebVitalsPayload: CoreWebVitalsPayload = {
+	browser_id: null,
+	page_view_id: null,
+	fid: null,
+	cls: null,
+	lcp: null,
+	fcp: null,
+	ttfb: null,
+};
+
+// By default, sample 1% of page views
+const userInSample = Math.random() < 1 / 100;
+
+let shouldSendMetrics = window.location.hash === '#forceSendMetrics';
+// unless we are forcing sending metrics for this page view
+// via initialisation or calling forceSendMetrics()
+export const forceSendMetrics = (): void => {
+	shouldSendMetrics = true;
+};
+
+const roundWithDecimals = (value: number, precision = 6): number => {
+	const power = Math.pow(10, precision);
+	return Math.round(value * power) / power;
+};
+
+const sendData = (): boolean => {
+	if (
+		!(userInSample || shouldSendMetrics) ||
+		coreWebVitalsPayload.fcp === null
+	)
+		return false;
+
+	const endpoint =
+		window.location.hostname === 'm.code.dev-theguardian.com' ||
+		window.location.hostname === 'localhost' ||
+		window.location.hostname === 'preview.gutools.co.uk'
+			? 'https://performance-events.code.dev-guardianapis.com/core-web-vitals'
+			: 'https://performance-events.guardianapis.com/core-web-vitals';
+
+	return navigator.sendBeacon(endpoint, JSON.stringify(coreWebVitalsPayload));
```

Comment:
Have we decided not to make the change from `sendBeacon` to `fetch`? I thought I understood that `fetch` had wider support.

Reply:
We’ve added a polyfill for `navigator.sendBeacon` on [frontend] and [DCR].

[frontend]: https://github.com/guardian/frontend/pull/24164
[DCR]: https://github.com/guardian/dotcom-rendering/pull/3415